torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -273,7 +273,8 @@ at::Tensor AtenXlaType::mul(const at::Tensor& self,
 
 at::Tensor& AtenXlaType::mul_(at::Tensor& self, const at::Tensor& other) const {
   XLATensor self_tensor = bridge::GetXlaTensor(self);
-  XLATensor::mul_(self_tensor, bridge::GetXlaTensor(other));
+  XLATensor::mul_(self_tensor,
+                  bridge::GetOrCreateXlaTensor(other, self_tensor.GetDevice()));
```

Comment:
This is weird, it means we effectively have to support a pair of device (`self`) and CPU (`other`) tensors for this operator. I don't know whether this is the norm for CUDA tensors, for example. Does this happen at steady state or is it a one-off to initialize buffers in the optimizer?

Reply:
Hopefully not everywhere, otherwise we will have to get rid if these three APIs and have a single one which always returns an XLATensor.
This is coming from optimizers, which we need to handle if we don't plan to patch them.
And patch them is not worky, is the grand plan of having a seamless integration ;)
The other thing that happens, they might come in as Double tensors.

Reply:
This is fine if it cannot be avoided and we have a good understanding why we end up with combinations of XLA and CPU, which is why I'm bringing up CUDA tensors. I'm not aware of any such mechanisms for those though, which makes me slightly concerned. Regarding them coming in as Double tensors, that shouldn't happen either if we run `torch.set_default_tensor_type('torch.FloatTensor')` before training the model.

It's not a problem _per se_ to do this, but I'd like to understand the implications. Especially the steady state performance implications -- having transfers during each optimizer step wouldn't be great.

Reply:
Well CUDA must have a smoother handling of curve balls than we have.
This is coming from optimizers state, which is no guarantee at all if it "on device".
Actually, it is mostly not. There were people asking for a optimizer.to() thing, which I have as well in my test_operations.py hack, but that does not help.
Also, we cannot expect people to call set_default_tensor_type(), which we do, and nevertheless we get Double tensors.

Reply:
Calling `set_default_tensor_type` can be done in the init, we can deal with that. Getting Double tensors back sounds like a bug to me. For SGD, what would we end up transferring each step? We are able to do this with no redundant transfers using `XLATensor` for the current trace-level integration, it'd be weird if we can't do it with `AtenXlaType`.

Reply:
We do set_default_tensor_type() in our test_operations.py and nevertheless we get Double tensors.
Besides, nothing prevents a user to create a pytorch.float64 tensor.

The performance issue is next to be addressed, there is a lot of stuff to be figured out there, WRT parallel transfers, graph constancy, and so on.